### DISCUSSION TITLE
Changes to extra-fields

### DISCUSSION THREAD

**Discussion by syfgk** [mortii/anki-morphs#221] [Ideas]
### Describe the bug

[//]: # (Please provide a clear and concise description of what the bug is.)
(For context, I'm using AnkiMorphs with the Chinese SpaCy models.)

Here's the bug: after running re-calc, some AnkiMorphs fields are not updating.

Below is one example. Notice that `am-unknowns` and `am-unknowns-count` are populated. However, I'd only expect this for new cards. This card, though, is one I've reviewed four times already. Because it's seen, I'd expect `am-unknowns` to be null and `am-unknowns-count` to be zero. Am I misunderstanding how re-calc works?

<img width="461" alt="image" src="https://github.com/mortii/anki-morphs/assets/164821375/aa8b3738-4e00-4479-af62-55a33bbd4f27">

<br>

<img width="336" alt="image" src="https://github.com/mortii/anki-morphs/assets/164821375/7545273b-18d5-4e3d-8976-9e41bce70b59">

I should note that this is not affecting all seen cards. There are other (seen) cards that are being re-calc'd as I'd expect (i.e., `am-unknowns` and `am-unknowns-count` are empty and 0, respectively).

### My setup

 - Operating System: MacOS
 - Anki Version:  Version ⁨23.12.1 (1a1d4d54)⁩; Python 3.9.15 Qt 6.5.3 PyQt 6.5.3
 - AnkiMorphs Version: v2.2.2

**Comment by mortii** [comment]
> Here's the bug: after running re-calc, some AnkiMorphs fields are **_not_** updating.

This seems to directly contradict this:

> Notice that am-unknowns and am-unknowns-count **_are populated_**. However, I'd only expect this for new cards.

I'll assume you meant that the extra fields are updating when the shouldn't.

The only extra field that should update on non-new cards is the `am-highlighted` field. In this context "new" means cards not in the new-queue, it might be possible to move cards to the new-queue if you reset them, and you might see the still see the review count if you don't select `Reset repetition and lapse count`, but I haven't checked.

You could double check if the fields are actually updating by erasing all the data in the extra fields and then running recalc again.

**Comment by syfgk** [comment]
Can we walk through a hypothetical example to confirm my understanding of how re-calc works?

Let's say, for instance, I have one new card which has 1 unseen morph. Since this is a card I have not reviewed yet, I would expect that running re-calc would result in the following:
- `am-unknowns` = whatever the 1 unseen morph is
- `am-unknowns-count` = 1

Now, _after_ I review the card, if I run re-calc again, my understanding is that I would now see the following:
- `am-unknowns` = blank
- `am-unknowns-count` = 0

Is my understanding correct?

**Comment by mortii** [comment]
> Now, _after_ I review the card, if I run re-calc again, my understanding is that I would now see the following:
> 
>     * `am-unknowns` = blank
> 
>     * `am-unknowns-count` = 0
> 
> 
> Is my understanding correct?

No, they will keep having the same values as when the card was new, since they don't get updated, in this case:
- `am-unknowns` = whatever
- `am-unknowns-count` = 1

This essentially freezes the card in the same state as when you first review it, that way you can have the `am-unknowns` field visible on your card and not lose it the next time you review. 

So for this card:
![image](https://github.com/mortii/anki-morphs/assets/15674619/a175684e-16e5-4aaa-a9fa-1132e0fc3c25)

I would still see the word on the top, even though it's not technically "unknown" to me anymore. This makes sure that useful context is not lost on future reviews.

**Comment by syfgk** [comment]
Ah, ok. So, follow-up question: why do some cards have:
- am-unknowns = blank
- am-unknowns-count = 0

**Comment by mortii** [comment]
If you have two cards in this order:

|         | card queue status | unknown morphs | unknown morphs count |
|---------|-------------------|----------------|----------------------|
| Card 01 | new               | whatever       | 1                    |
| Card 02 | new               | whatever       | 1                    |


and you learn the first card, then the morphs on that first card will no longer be unknown on any new cards:

|         | card queue status | unknown morphs | unknown morphs count |
|---------|-------------------|----------------|----------------------|
| Card 01 | review            | whatever       | 1                    |
| Card 02 | new               |                | 0                    |


does that make sense?

**Comment by syfgk** [comment]
It's making more sense! In this case we're assuming the "unknown morphs" are the same across both cards, right?

**Comment by mortii** [comment]
> In this case we're assuming the "unknown morphs" are the same across both cards, right?

Sorry, it wasn't a great example, but yes, both cards have the same unknown morph "whatever". Substitute it with "airplaine" or "人" or any other morph.

Btw, I'm not saying this current approach is good, in fact, the `am-unknowns-count` field should probably update for all cards regardless if they are new or not, that way the sorting by that field would be more useful. Do you have any suggestions/preferences for a better approach?

**Comment by syfgk** [comment]
Ah! I see, ok things are making sense now. So let me ask this:

Let's say I have these cards, both are new:

|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | new               | A, B           | 2                    |
| Card 2 | new               | A              | 1                    |

Now let's say I review the first card. My understanding is that this is what I'd now see:

|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | review               | A, B           | 2                    |
| Card 2 | new               |                | 0                    |

Is this correct?

Edit: Thinking about this again, I guess the wrench here is that I'd technically see Card 2 first, right? Since that's the one with the fewest morphs. So my tables should instead be something like:

Table 1 (both cards are new)
|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | new               | A              | 1                    |
| Card 2 | new               | A,B            | 2                    |

Table 2 (after `Card 1` has been reviewed):
|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | review               | A              | 1                    |
| Card 2 | new               | B            | 1                   |

**Comment by mortii** [comment]
Yep, both of those scenarios are correct, although the first one should realistically only happen during some custom review circumstances or if they are in different decks.

**Comment by syfgk** [comment]
Nice! Here's my suggestion: have _**four**_ extra fields instead of two

- `unknown-morphs`: shows all the unknown/unseen morphs on a new card. Once a card has been reviewed/seen and re-calc is run, this field goes blank

- `unknown-morphs-count`: displays the count of `unknown-morphs`. Once a card has been reviewed/seen and re-calc is run, this field shows `0`

- _**(new)**_ `all-morphs`: this shows all morphs on a card, regardless of (a) known/unknown, or (b) card state. So even after the user has reviewed/seen the card, this field stays the same

- _**(new)**_ `all-morphs-count`: this is a count of `all-morphs`; same logic applies. After the user has reviewed/seen the card, this field also stays the same

Here’s an example:

Table 1 (card is new)

|        | card queue status | unknown morphs | unknown morphs count | all morphs | all morphs count |
|--------|-------------------|----------------|----------------------|------------|------------------|
| Card 1 | new               | A,B,C,D        | 4                    | A,B,C,D    | 4                |

Table 2 (card has been reviewed)

|        | card queue status | unknown morphs | unknown morphs count | all morphs | all morphs count |
|--------|-------------------|----------------|----------------------|------------|------------------|
| Card 1 | review               |         | 0                    | A,B,C,D    | 4                |

What do you think?

**Comment by mortii** [comment]
That is very logical, and I don't mind adding new extra fields (I coincidentally reached the same conclusion about adding an "all morphs count" extra field in #195)

I don't really see how prevent the loss of the unknown morph after the first review though, like I mentioned in my previous comment https://github.com/mortii/anki-morphs/issues/220#issuecomment-2053691213, that context can be pretty important later. Any ideas?

**Comment by syfgk** [comment]
Is the image in that comment both sides of your card, or just one side? I want to make sure I'm oriented correctly and know what's on the front and back.

**Comment by mortii** [comment]
Here is another example card:

![Screenshot from 2024-04-14 15-14-34](https://github.com/mortii/anki-morphs/assets/15674619/15e8790f-dbb6-47de-9cce-c0189d46261d)
![Screenshot from 2024-04-14 15-15-00](https://github.com/mortii/anki-morphs/assets/15674619/d08c39cc-3d27-4d35-aada-75d50c3eac28)

the template looks something like this (simplified):
![image](https://github.com/mortii/anki-morphs/assets/15674619/4f6078f5-aca9-4ab6-b853-2829234c6646)

where the `am-unknowns` extra field is show on the front and back.

**Comment by syfgk** [comment]
Thanks for sharing! What's the value of showing `am-unknowns` on the front instead of just having `am-highlighted`?

**Reply by mortii** [reply]
I converted this to a discussion since the scope changed and the reply feature is better. 

Anyway, having the `am-unknowns` field on the cards is somewhat of a remnant of the pre `am-highlighted` era, but it does serve the purpose of letting me know what the unknown morph was at the time when I first studied the card. `am-highlighted` always updates, so that particular morph will have the same color as the other morphs eventually, i.e. that information gets lost.

**Reply by syfgk** [reply]
All good; thanks for the explanation!

One idea is: can you collapse `am-unknowns` and `am-highlighted`? Basically keep `am-highlighted` static?

I struggle to see the value of keeping `am-highlighted` as it currently is, since it's essentially redundant. At some point, it'll end up looking the same as whatever field the user is using as their note filter field.

Edit: By keep `am-highlighted` static, I mean: highlight whatever morphs were unknown at the time of first review, and then keep those highlights thereafter.

**Reply by mortii** [reply]
> I struggle to see the value of keeping `am-highlighted` as it currently is, since it's essentially redundant. At some point, it'll end up looking the same as whatever field the user is using as their note filter field.

I would completely agree with this if the morphs in `am-highlighted` only toggled between "unknown" and "known", but they also change color when they reach the "learning" state, so it's an interesting progression feedback mechanism. 

It also occurred to me that the [context menu option "browse in am-unknowns"](https://mortii.github.io/anki-morphs/user_guide/usage/reviewing-cards.html#right-clicking-highlighted-text) that @xofm31 made, would stop working if that field is still dynamic after the first study.

So, as far as I see it, there are two options:
1. break backwards compatibility and accept that some context will be lost after the first study
2. add an additional "archival" extra field, something like "unknown-when-first-studied"

but maybe I'm not thinking outside the box enough. 

Thoughts?

**Reply by syfgk** [reply]
> I would completely agree with this if the morphs in am-highlighted only toggled between "unknown" and "known", but they also change color when they reach the "learning" state, so it's an interesting progression feedback mechanism.

Oh that's interesting; I didn't know about that. Hm; I feel like there are a lot of moving pieces here...

Ok, maybe let's work backwards from your two solutions and go from there.

> break backwards compatibility and accept that some context will be lost after the first study

I don't want to assume I know what all this solution entails. Do you mind elaborating? What exactly would be changing with this solution?

> add an additional "archival" extra field, something like "unknown-when-first-studied"

Same thing here. How would the experience change for the user?

> It also occurred to me that the [context menu option "browse in am-unknowns"](https://mortii.github.io/anki-morphs/user_guide/usage/reviewing-cards.html#right-clicking-highlighted-text) that @xofm31 made, would stop working if that field is still dynamic after the first study.

I didn't know about this either, but I'm wondering if I'm misunderstanding how this feature works. Looking at the help guide description, it seems this feature culls through all `am-unknown` fields and looks for cards that have this morph? But per our discussion thus far, wouldn't we expect the search results to only return one card? Per this [comment](https://github.com/mortii/anki-morphs/discussions/221#discussioncomment-9109345), once a new morph is seen, that morph gets wiped from the `am-unknown` field of all other cards, right?

**Comment by mortii** [comment]
Okay, so the two options I have in mind are:
- Only changing how the current extra fields work
- Changing how the extra fields work **and** adding an archival field

### Only changing how the current extra fields work

This is a straight forward implementation of your suggestions:

>- `unknown-morphs`: shows all the unknown/unseen morphs on a new card. Once a card has been reviewed/seen and re-calc is run, this field goes blank
>
>- `unknown-morphs-count`: displays the count of `unknown-morphs`. Once a card has been reviewed/seen and re-calc is run, this field shows `0`

this is a completely valid approach, but since it changes how they currently work it will break backwards compatibility in these ways:
- the unknown morph will disappear from the card after studying it for the first time, which will break the setup some people already have.
- since the unknown morph disappears, the `browse "am-unknowns"` feature will no longer find  the cards it does currently. I could go into more detail on how this feature works, but for our purposes, it might be sufficient to just say it will no longer work as advertised.

We could just conclude that this approach is so much better than the current one that we just implement it, don't worry about any backwards compatibility issues, and move on.

### Changing how the extra fields work **and** adding an archival field

This would involve making the same changes to the extra fields as the option above:

>- `unknown-morphs`: shows all the unknown/unseen morphs on a new card. Once a card has been reviewed/seen and re-calc is run, this field goes blank
>
>- `unknown-morphs-count`: displays the count of `unknown-morphs`. Once a card has been reviewed/seen and re-calc is run, this field shows `0`

but we also add the extra field `am-unknowns-when-first-studied` that will have the same logic as the current `am-unknowns`.  People could replace the `am-unknowns` field on their card templates with `am-unknowns-when-first-studied`, and then that information won't be lost. 

The same thing could be done with the context menu option:
`browse in am-unknowns` -> `browse in am-unknowns-when-first-studied`


Both of these break backwards compatibility, but the second option just mitigates it somewhat, so these changes would warrant upgrading the AnkiMorphs version to v3.

**Reply by syfgk** [reply]
Thanks for writing this out! I'd vote for # 1 since it feels more intuitive.

As a side note, I'm kinda curious: does Anki/AnkiMorphs track any analytics on how folks are actually using the add-on? I feel like some of this decision-making could be made much easier if it was possible to see what features were being used/not used and how (in)frequently.

**Reply by mortii** [reply]
> does Anki/AnkiMorphs track any analytics on how folks are actually using the add-on?

AnkiMorphs definitely doesn't track any user behavior, nor would I ever want to open that can of worms. I doubt Anki does anything like that either, but I don't actually know.

**Reply by syfgk** [reply]
Good to know!

And thanks again for investing in this project. It's been great working with you to nail this down.

**Comment by xofm31** [comment]
I think I now understand why Morphman has a separate field for Focus Morph (required) and Unknowns. This way Focus Morph isn't updated once the card isn't new anymore and Unknowns continues to be updated.

My setup is similar to @mortii's where I have the "Focus Morph" am-unknowns at the top of my card. My primary goal in using Ankimorphs / Morphman is to order the vocabulary that Anki introduces. Here is an example card:

<img width="392" alt="Screenshot " src="https://github.com/mortii/anki-morphs/assets/14337605/2014caa4-a9c9-4123-8583-450497b60d50">

The sentence at the bottom is the field listed in Ankimorphs settings. If it came with translation / pinyin, I use that; if not, I use Google Translate.

The word at the top is am-unknowns. If it exists in the cc-cedict, it automatically populates the pinyin and definition. If not, I look it up and manually enter it.

When I am studying, the most important is the word. I want to learn the word, not the sentence, so I grade the card only based on the am-unknowns field. Please don't get rid of this field! (or make two fields like Morphman has). I don't see any reason to keep `am-unknowns-count` at 1 once the am-unknowns word has been learned though.

I generate the highlighting of the word in the sentence in the card:
```
<span id="Sentence" class="sentence"> {{Sentence}} </span>
<script> 
var w = '{{Word}}'; 
var s = document.getElementById("Sentence").innerHTML; document.getElementById("Sentence").innerHTML = s.replaceAll(w, "<h class=word>" + w + "</h>"); 
</script>
```

**Reply by mortii** [reply]
> Please don't get rid of this field! (or make two fields like Morphman has).

Do you mean that we should make it into two fields like morphman? That is really the only viable path forward in my opinion.

**Reply by xofm31** [reply]
If you really want to have a field that shows the morpheme(s) that are still unknown, then yes, two fields.

**Reply by mortii** [reply]
How about naming it `am-first-study-unknowns`? I really dislike "focus morph" because it is not descriptive at all.

**Reply by xofm31** [reply]
I have no preference on the name.

**Reply by xofm31** [reply]
Since you are talking about these fields, I will point you to my code where I am trying to alter the way that the morph interval is calculated. I think I said before that I would like it to use the interval of the card(s) with that morph as the "focus morph" if there is a card with that as the focus morph, and if not, then use the interval of any card with that morph. I have it working for me, but only because I added a new field "focus_morph" to the card. I thought that your rework of the extra fields might help me, but I still don't see an easy way to do this, because I either need that focus morph in the card_data, or I need access to the note.

As you're working through the field changes, maybe you can keep this in mind (or maybe I have missed an easy way to do this...)

https://github.com/xofm31/anki-morphs/blob/40b200eb1ae96650da7beffb5e839e87134936f5/ankimorphs/recalc.py#L311-L323

Also, if you're planning to break backwards compatibility anyway, you may want to consider changing the field name "field" in the ankimorphs_profile_settings.json to something more meaningful and distinctive.

**Reply by mortii** [reply]
@xofm31 I've added responses to those in #181 and #224 respectively

**Comment by xofm31** [comment]
> I didn't know about this either, but I'm wondering if I'm misunderstanding how this feature works. Looking at the help guide description, it seems this feature culls through all am-unknown fields and looks for cards that have this morph? But per our discussion thus far, wouldn't we expect the search results to only return one card? Per this https://github.com/mortii/anki-morphs/discussions/221#discussioncomment-9109345, once a new morph is seen, that morph gets wiped from the am-unknown field of all other cards, right?

The main use case for this is for looking up words in the sentence. From the card above, I have the sentence: 我今天好丢脸. The focus morph on the card is 丢脸, but if I forget 今天, I can look it up this way. Also, there are a lot of words that have more than one meaning, so I may have multiple cards for the same focus morph, and when I come across it, I may want to find all the cards with that focus morph to remember the multiple meanings. (As I'm watching a show, I come across a word that I "know", but it doesn't make sense, so I look it up, and realize that it's got a different meaning. So I find that card which has moved to the bottom of the queue, manually put in the am-unknowns, "AnkiMoprhs....Learn Card Now" and learn it before recalculating).

**Comment by xofm31** [comment]
Since you are reworking so many things, I had another idea, to deal with mistakes in the morphemizer. 

> Stuttered names or words might accidentally produce morphs that don't make any sense in the context, and you should probably suspend these cards or mark them as known if there are many of them.

Sometimes the morphs don't make sense in the context, but sometimes they just aren't actually real "words" - it splits up two words and combines the last character of one and the first character of the next into something that is not a word.  Especially since the sentence will still be in the text that I'm using for my study plan, I'd rather have a way to mark the morph as "invalid" or "ignore" rather than suspending or deleting individual cards. Right now I'm seeing what happens if I mark them as "Names" and selecting "Ignore names found in names.txt" but this doesn't seem like a satisfactory solution.

**Reply by mortii** [reply]
Interesting. I created a new discussion for it: #223

**Comment by mortii** [comment]
Released in v3: https://github.com/mortii/anki-morphs/releases/tag/v3.0.0

Thanks!
